File: libmuscle/python/libmuscle/mcp/tcp_transport_client.py
import select
import logging
import socket
import time
from typing import Optional, Tuple

from libmuscle.mcp.transport_client import ProfileData, TransportClient, TimeoutHandler
from libmuscle.mcp.tcp_util import (
        is_disconnect, recv_frame, recv_int64, send_frame, send_int64)
from libmuscle.profiling import ProfileTimestamp
from libmuscle.util import Retrier


_logger = logging.getLogger(__name__)


_CONNECT_TIMEOUT = 3.0                      # seconds
_CONNECT_TIMEOUT_PATIENT = 60.0             # seconds
_CONNECT_TIMEOUT_PATIENT_STEP = 3.0         # seconds
_RECONNECT_TIMEOUT = 60.0                   # seconds


class NoPendingResponse(RuntimeError):
    pass


class TcpTransportClient(TransportClient):
    """A client that connects to a TCPTransport server."""
    @staticmethod
    def can_connect_to(location: str) -> bool:
        """Whether this client class can connect to the given location.

        Args:
            location: The location to potentially connect to.

        Returns:
            True iff this class can connect to this location.
        """
        return location.startswith('tcp:')

    def __init__(self, location: str) -> None:
        """Create a TcpClient for a given location.

        The client will connect to this location and be able to send requests to it and
        return the response.

        Args:
            location: A location string for the peer.
        """
        self._addresses = location[4:].split(',')
        self._socket: Optional[socket.SocketType] = None
        self._session = 0
        self._cur_request = 0

        self._reconnect(False)

    def call(self, request: bytes, timeout_handler: Optional[TimeoutHandler] = None
             ) -> Tuple[bytes, ProfileData]:
        """Send a request to the server and receive the response.

        This is a blocking call.

        Args:
            request: The request to send
            timeout_handler: Optional timeout handler. This is used for communication
                deadlock detection.

        Returns:
            The received response
        """
        self._cur_request += 1
        retrier = Retrier(_RECONNECT_TIMEOUT)
        while True:
            try:
                start_wait = ProfileTimestamp()
                if self._socket is None:
                    raise ConnectionError('No connection could be established')

                send_int64(self._socket, self._cur_request)
                send_frame(self._socket, request)

                did_timeout = False
                if timeout_handler is not None:
                    while not self._poll(timeout_handler.timeout):
                        did_timeout = True
                        timeout_handler.on_timeout()

                if did_timeout:
                    assert timeout_handler is not None  # mypy
                    timeout_handler.on_receive()

                start_transfer = ProfileTimestamp()
                response = recv_frame(self._socket)
                stop_transfer = ProfileTimestamp()
                return response, (start_wait, start_transfer, stop_transfer)

            except Exception as e:
                if is_disconnect(e):
                    self._handle_disconnect(retrier)
                else:
                    raise

    def close(self) -> None:
        """Closes this client.

        This closes any connections this client has and performs other shutdown
        activities as needed.
        """
        try:
            if self._socket is not None:
                self._socket.shutdown(socket.SHUT_RDWR)
                self._socket.close()
        except Exception as e:
            # This can raise if the peer has shut down already when we close our
            # connection to it, which is fine and can be ignored. Otherwise, we reraise.
            if not is_disconnect(e):
                raise

    def _poll(self, timeout: float) -> bool:
        """Poll the socket and return whether its ready for receiving.

        This method blocks until the socket is ready for receiving, or :param:`timeout`
        seconds have passed (whichever is earlier).

        Args:
            timeout: timeout in seconds

        Returns:
            True if the socket is ready for receiving data, False otherwise.
        """
        retrier = Retrier()
        while True:
            try:
                if self._poll_obj is not None:
                    ready = self._poll_obj.poll(timeout * 1000)  # poll timeout is in ms
                else:
                    # Fallback to select()
                    ready, _, _ = select.select([self._socket], (), (), timeout)
                return bool(ready)

            except Exception as e:
                if is_disconnect(e):
                    self._handle_disconnect(retrier)
                else:
                    raise

    def _handle_disconnect(self, retrier: Retrier) -> None:
        """Handles a broken network connection.

        Args:
            retrier: A Retrier that keeps track of timing any retries
        """
        _logger.warning(
                f'The TCP network connection with {self._addresses} was lost'
                ' unexpectedly.')

        try:
            self.close()
        except Exception as e:
            if not is_disconnect(e):
                raise

        if retrier.should_give_up():
            _logger.warning(
                    f'I am unable to reconnect to {self._addresses} despite repeated'
                    ' attempts, and I am giving up. Please check your network.')
            raise

        retrier.sleep()

        _logger.warning(f'Trying to reconnect to {self._addresses}')

        self._reconnect()

    def _reconnect(self, re: bool = True) -> None:
        """(Re)connect to the server and resume the current session

        Args:
            re: True if this is a reconnect rather than an initial connect.
        """
        try:
            self._make_connection()
            assert self._socket is not None
            send_int64(self._socket, self._session)
            self._session = recv_int64(self._socket)

            if re:
                _logger.warning(
                        f'Reconnected to {self._addresses}, continuing the'
                        ' simulation')

        except Exception as e:
            if is_disconnect(e):
                self.close()
                _logger.warning(
                        f'Failed to reconnect to {self._addresses}, will retry'
                        ' later')
            else:
                raise

    def _make_connection(self) -> None:
        """Connect to the server and set up polling

        Uses self._addresses and creates a (new) self._socket and self._poll_obj.
        """
        sock: Optional[socket.SocketType] = None
        for address in self._addresses:
            try:
                sock = self._connect(address, False)
                break
            except RuntimeError:
                pass

        if sock is None:
            # None of our quick connection attempts worked. Either there's a network
            # problem, or the server is very busy. Let's try again with more patience.
            _logger.warning(
                    f'Could not immediately connect to {self._addresses}, trying again'
                    ' with more patience. Please report this if it happens frequently.')

            for address in self._addresses:
                try:
                    _logger.debug(
                            f'Trying to connect to {address} patiently')
                    sock = self._connect(address, True)
                    break
                except RuntimeError:
                    pass

        if sock is None:
            _logger.error(
                    f'Failed to connect also on the second try to {self._addresses}')
            raise ConnectionRefusedError(
                    f'Could not connect to any server at locations {self._addresses}')

        if hasattr(socket, "TCP_NODELAY"):
            sock.setsockopt(socket.SOL_TCP, socket.TCP_NODELAY, 1)
        if hasattr(socket, "TCP_QUICKACK"):
            sock.setsockopt(socket.SOL_TCP, socket.TCP_QUICKACK, 1)
        self._socket = sock

        if hasattr(select, "poll"):
            self._poll_obj: Optional[select.poll] = select.poll()
            self._poll_obj.register(self._socket, select.POLLIN)
        else:
            self._poll_obj = None  # On platforms that don't support select.poll

    def _connect(self, address: str, patient: bool) -> socket.SocketType:
        loc_parts = address.rsplit(':', 1)
        host = loc_parts[0]
        if host.startswith('['):
            if host.endswith(']'):
                host = host[1:-1]
            else:
                raise RuntimeError('Invalid address')
        port = int(loc_parts[1])

        timeout = _CONNECT_TIMEOUT_PATIENT if patient else _CONNECT_TIMEOUT
        patient_step = _CONNECT_TIMEOUT_PATIENT_STEP

        addrinfo = socket.getaddrinfo(
                host, port, 0, socket.SOCK_STREAM, socket.IPPROTO_TCP)

        for family, socktype, proto, _, sockaddr in addrinfo:
            try:
                sock = socket.socket(family, socktype, proto)
            except Exception:
                continue

            start_time = time.monotonic()
            time_left = (start_time + timeout) - time.monotonic()
            while (time_left > 0.0):
                try:
                    sock.settimeout(time_left)
                    sock.connect(sockaddr)
                    sock.settimeout(None)
                    return sock
                except (ConnectionRefusedError, ConnectionAbortedError):
                    if patient:
                        _logger.warning('Connection refused, sleeping')
                        time.sleep(patient_step)
                    else:
                        _logger.info(f'Failed to connect to {sockaddr}')
                        sock.close()
                        break

                except Exception as e:
                    _logger.debug(f'Failed to connect socket: {e}')
                    sock.close()
                    break

                time_left = (start_time + timeout) - time.monotonic()

        raise RuntimeError('Could not connect')

File: libmuscle/cpp/src/libmuscle/mcp/tcp_transport_client.cpp
#include "libmuscle/mcp/tcp_transport_client.hpp"

#include <libmuscle/logger.hpp>
#include <libmuscle/mcp/tcp_util.hpp>

#include <algorithm>
#include <cassert>
#include <cstring>
#include <chrono>
#include <cmath>
#include <exception>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>
#include <unistd.h>
#include <vector>

#include <sys/types.h>
#include <sys/socket.h>
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/tcp.h>
#include <poll.h>
#include <string.h>
#include <unistd.h>


namespace {

using libmuscle::_MUSCLE_IMPL_NS::log_debug;
using libmuscle::_MUSCLE_IMPL_NS::log_info;
using libmuscle::_MUSCLE_IMPL_NS::log_warning;
using libmuscle::_MUSCLE_IMPL_NS::time_monotonic;
using libmuscle::_MUSCLE_IMPL_NS::mcp::check_conn;
using libmuscle::_MUSCLE_IMPL_NS::mcp::ConnectionRefused;
using libmuscle::_MUSCLE_IMPL_NS::mcp::do_poll_out;
using std::chrono::duration;


const double connect_timeout = 3.0;                 // seconds
const double reconnect_timeout = 60.0;              // seconds


/* Splits a location string of the form tcp:<address:port>,<address:port>,...
 * into a list of addresses.
 */
std::vector<std::string> split_location(std::string const & location) {
    std::vector<std::string> addresses;

    // start at 4 to skip the initial tcp: bit
    for (auto it = std::next(location.cbegin(), 4); it != location.cend(); ) {
        auto next = std::find(it, location.cend(), ',');
        addresses.emplace_back(it, next);

        it = next;
        if (it != location.cend())
            ++it;
    }

    return addresses;
}


int connect(std::string const & address) {
    std::string errors;

    std::size_t split = address.rfind(':');
    std::string host = address.substr(0, split);
    if (host.front() == '[') {
        if (host.back() == ']')
            host = host.substr(1, host.size() - 2);
        else
            throw std::runtime_error("Invalid address");
    }
    std::string port = address.substr(split + 1);

    int err_code;

    // collect addresses to connect to
    addrinfo hints;
    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    addrinfo *res;
    err_code = getaddrinfo(host.c_str(), port.c_str(), &hints, &res);
    if (err_code != 0)
        throw std::runtime_error("Could not connect to " + host + " on port "
                + port + ": " + gai_strerror(err_code));
    auto address_info = std::unique_ptr<addrinfo, void (*)(addrinfo*)>(res, &freeaddrinfo);

    // try to connect to each in turn until we find one that works
    addrinfo * p;
    for (p = address_info.get(); p != nullptr; p = p->ai_next) {
        int socket_fd = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
        if (socket_fd == -1) continue;

        const double start_time = time_monotonic();
        double time_left = (start_time + connect_timeout) - time_monotonic();

        while (time_left > 0.0) {
            try {
                int flags = fcntl(socket_fd, F_GETFL, 0);
                fcntl(socket_fd, F_SETFL, flags | O_NONBLOCK);

                connect(socket_fd, p->ai_addr, p->ai_addrlen);
                do_poll_out(socket_fd, time_left);

                // check if connect() actually succeeded
                socklen_t len = sizeof(int);
                getsockopt(socket_fd, SOL_SOCKET, SO_ERROR, &err_code, &len);
                check_conn(err_code);

                flags = fcntl(socket_fd, F_GETFL, 0);
                fcntl(socket_fd, F_SETFL, flags & ~O_NONBLOCK);

                return socket_fd;
            }
            catch (ConnectionRefused const & e) {
                log_info("Failed to connect to ", host);
                ::close(socket_fd);
                break;
            }
            catch (std::exception const & e) {
                log_debug("Failed to connect socket ", e.what());
                ::close(socket_fd);
                break;
            }

            time_left = (start_time + connect_timeout) - time_monotonic();
        }
    }

    throw ConnectionRefused("Could not connect");
}

}


namespace libmuscle { namespace _MUSCLE_IMPL_NS { namespace mcp {

bool TcpTransportClient::can_connect_to(std::string const & location) {
    return location.compare(0u, 4u, "tcp:") == 0;
}

TcpTransportClient::TcpTransportClient(std::string const & location)
    : location_(location)
    , socket_fd_(-1)
    , session_(0)
    , cur_request_(0)
{
    addresses_ = split_location(location);
    reconnect_(false);
}

TcpTransportClient::~TcpTransportClient() {
    if (socket_fd_ != -1) {
        end_session_();
        close_connection_();
    }
}

std::tuple<std::vector<char>, ProfileData> TcpTransportClient::call(
        char const * req_buf, std::size_t req_len, TimeoutHandler* timeout_handler
) {
    ++cur_request_;
    Retrier retrier(reconnect_timeout);
    double deadline = std::numeric_limits<double>::infinity();
    bool did_timeout = false;

    while (true) {
        try {
            if (deadline < time_monotonic()) {
                timeout_handler->on_timeout();
                deadline += timeout_handler->get_timeout();
                did_timeout = true;
            }

            if (socket_fd_ == -1)
                throw Disconnect("No connection could be established");

            ProfileTimestamp start_wait;
            send_request_(cur_request_, req_buf, req_len);

            if (timeout_handler != nullptr) {
                if (std::isinf(deadline))
                    deadline = time_monotonic() + timeout_handler->get_timeout();

                pollfd socket_poll_fd;
                socket_poll_fd.fd = socket_fd_;
                socket_poll_fd.events = POLLIN;
                while (poll_retry_eintr(
                            &socket_poll_fd, 1, deadline - time_monotonic()) == 0) {
                    timeout_handler->on_timeout();
                    deadline += timeout_handler->get_timeout();
                    did_timeout = true;
                }

                if (did_timeout) {
                    timeout_handler->on_receive();
                    did_timeout = false;
                }
            }

            ProfileTimestamp start_transfer;
            std::vector<char> result = recv_frame(socket_fd_);
            ProfileTimestamp stop_transfer;
            return std::make_tuple(
                    std::move(result),
                    std::make_tuple(start_wait, start_transfer, stop_transfer));
        }
        catch (Disconnect const & e) {
            handle_disconnect_(retrier);
        }
    }
}

void TcpTransportClient::close() {
    end_session_();
    close_connection_();
}

/** Sends a request in a single send call.
 *
 * This is an optimisation, we could just send_int64() the request_nr and then
 * send_frame() the data, but that causes short messages to be split across two packets,
 * with a 40ms delay in between. So we combine them here.
 */
void TcpTransportClient::send_request_(
        int64_t request_nr, char const * req_buf, std::size_t req_len)
{
    static_assert(sizeof(ssize_t) == 8, "MUSCLE3 needs a 64-bit machine/OS to compile");

    const ssize_t length = static_cast<ssize_t>(req_len);

    assert(length + 16 < std::numeric_limits<ssize_t>::max());

    const char * req_nr_data = reinterpret_cast<char*>(&request_nr);
    const char * len_data = reinterpret_cast<const char*>(&length);

    std::vector<char> buf(length + 16);
    std::copy(req_nr_data, req_nr_data + 8, buf.data());
    std::copy(len_data, len_data + 8, buf.data() + 8);
    std::copy(req_buf, req_buf + length, buf.data() + 16);

    send_all(socket_fd_, buf.data(), length + 16);
}

/** Handles a broken network connection.
 *
 * @param retrier A Retries that keeps track of timing any retries
 */
void TcpTransportClient::handle_disconnect_(Retrier & retrier) {
    log_warning("The TCP network connection with ", location_, " was lost unexpectedly.");

    try {
        close_connection_();
    }
    catch (Disconnect const & e) {}

    if (retrier.should_give_up()) {
        log_warning(
                "I am unable to reconnect to ", location_, " despite repeated",
                " attempts, and I am giving up. Please check your network.");
        throw;
    }

    retrier.sleep();
    log_warning("Trying to reconnect to ", location_);
    reconnect_();
}


/** (Re)connect to the server and resume the current session
 *
 * @param re True if this is a reconnect rather than an initial connect
 */
void TcpTransportClient::reconnect_(bool re) {
    try {
        make_connection_();
        send_int64(socket_fd_, session_);
        session_ = recv_int64(socket_fd_);

        if (re) {
            log_warning("Reconnected to ", location_, ", continuing the simulation");
        }
    }
    catch (Disconnect const & e) {
        close_connection_();
        log_warning("Failed to reconnect to ", location_,  ", will retry later");
    }
}


/** Connect to the server
 *
 * Uses addresses_ and creates a new socket_fd_ if successful.
 */
void TcpTransportClient::make_connection_() {
    std::string errors;
    int sock_fd = -1;

    for (auto const & address: addresses_)
        try {
            sock_fd = connect(address);
            break;
        }
        catch (std::runtime_error const & e) {
            errors += std::string(e.what()) + "\n";
        }

    if (sock_fd == -1) {
        throw Disconnect("Failed to connect");
    }

    int flags = 0;
#ifdef __linux
    setsockopt(sock_fd, SOL_TCP, TCP_NODELAY, &flags, sizeof(flags));
    setsockopt(sock_fd, SOL_TCP, TCP_QUICKACK, &flags, sizeof(flags));
#elif __APPLE__
    setsockopt(sock_fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof(flags));
    // macOS doesn't have quickack unfortunately
#endif

    socket_fd_ = sock_fd;
}


/** Tell the server we want to end our session
 *
 * We ignore errors here and rely on the server to time out if we don't manage to send.
 */
void TcpTransportClient::end_session_() {
    // End session
    try {
        send_int64(socket_fd_, 0);
    }
    catch (Disconnect const &) {
        log_warning(
                "Disconnected while trying to end session, shutdown will take longer",
                " than usual because of this.");
    }
}


/** Close the network connection and reset the socket fd
 */
void TcpTransportClient::close_connection_() {
    ::close(socket_fd_);
    socket_fd_ = -1;
}


} } }

